--- a/meson.build
+++ b/meson.build
@@ -80,6 +82,7 @@
 prog_python = import('python').find_installation('python3')
 
 with_host_windows = host_machine.system() == 'windows'
+with_host_android = host_machine.system() == 'android'
 with_host_darwin = host_machine.system() == 'darwin'
 
 not_found = dependency('', required: false)
@@ -295,7 +298,7 @@
 
 if with_egl
    if cc.has_header('epoxy/egl.h', dependencies: epoxy_dep) and epoxy_dep.get_variable(pkgconfig: 'epoxy_has_egl') == '1'
-      if with_host_windows or with_host_darwin
+      if with_host_windows or with_host_darwin or with_host_android
          have_egl = true
       else
          if libdrm_dep.found()
